### PR TITLE
[DNM] add prewrite request

### PR DESCRIPTION
**Pull request by lishihai9017**
PTAL @zhangjinpeng1987 @BusyJay @siddontang

### CODE REVIEW

**Comment by zhangjinpeng1987** [comment]
Please add test

**Comment by zhangjinpeng1987** [comment]
Please fix ci

File: src/raftstore/store/worker/apply.rs [hunk 1]
```diff
@@ -972,6 +974,41 @@ impl ApplyDelegate {
     fn handle_snap(&mut self, _: &ExecContext, _: &Request) -> Result<Response> {
         do_snap(self.region.clone())
     }
+
+    fn handle_prewrite(&mut self, ctx: &ExecContext, req: &Request) -> Result<Response> {
+        let prewrite = req.get_prewrite();
```
Comment (siddontang) [inline src/raftstore/store/worker/apply.rs]: how can we use coprocessor to handle this? @BusyJay

**Comment by siddontang** [comment]
please add tests and benchmark.

**Comment by lishihai9017** [comment]
### sysbench benchmark result
#### before optimized
```
tps: 7392.31 per sec.
response time:
    min:                                 10.14ms
    avg:                                 34.51ms
    max:                               1178.30ms
    approx.  95 percentile:              46.31ms
```

#### after optimized
```
tps: 7418.23 per sec.
response time:
    min:                                  3.99ms
    avg:                                 34.63ms
    max:                                587.19ms
    approx.  95 percentile:              46.36ms
```

File: Cargo.toml [hunk 2]
```diff
@@ -83,6 +84,7 @@ features = ["nightly", "push", "process"]
 
 [dependencies.grpc]
 git = "https://github.com/stepancheg/grpc-rust.git"
+rev = "13bdc724e5d9467792b289df822c2479812a9808"
```
Comment (zhangjinpeng1987) [inline Cargo.toml]: Is it necessary to modify grpc?

File: src/raftstore/coprocessor/prewrite_observer.rs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+use super::{Coprocessor, RegionObserver, ObserverContext};
+use kvproto::raft_cmdpb::{RaftCmdRequest, CmdType, Request, PutRequest};
+use storage::types::Key;
+use storage::CF_LOCK;
+
+use protobuf::RepeatedField;
+
+pub struct PrewriteObserver;
+
+impl Coprocessor for PrewriteObserver {}
```
Comment (zhangjinpeng1987) [inline src/raftstore/coprocessor/prewrite_observer.rs]: s/PrewriteObserver/TxnObserver

File: src/raftstore/coprocessor/mod.rs [hunk 4]
```diff
@@ -61,12 +61,12 @@ pub trait RegionObserver: Coprocessor {
     }
 
     /// Hook to call before execute read/write request.
-    fn pre_query(&self, _: &mut ObserverContext, _: &mut RepeatedField<Request>) -> Result<()> {
+    fn pre_query(&self, _: &mut ObserverContext, _: &mut RaftCmdRequest) -> Result<()> {
```
Comment (BusyJay) [inline src/raftstore/coprocessor/mod.rs]: Why changes the interface?

File: src/raftstore/coprocessor/prewrite_observer.rs [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+impl RegionObserver for PrewriteObserver {
+    fn pre_apply_query(&self, ctx: &mut ObserverContext, cmd_req: &mut RaftCmdRequest) {
+        let mut new_reqs = Vec::<Request>::with_capacity(cmd_req.get_requests().len());
+        for req in cmd_req.get_requests() {
+            match req.get_cmd_type() {
```
Comment (BusyJay) [inline src/raftstore/coprocessor/prewrite_observer.rs]: Use if to reduce indention.

File: src/raftstore/coprocessor/prewrite_observer.rs [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+                    new_req.set_put(put);
+                    new_reqs.push(new_req.to_owned());
+                }
+                _ => {
+                    new_reqs.push(req.to_owned());
+                }
+            }
+        }
+        cmd_req.set_requests(RepeatedField::from_vec(new_reqs));
+        ctx.bypass = true;
```
Comment (BusyJay) [inline src/raftstore/coprocessor/prewrite_observer.rs]: Why bypass?

**Comment by lishihai9017** [comment]
#### bench result
`test raftstore::coprocessor::prewrite_observer::tests::bench_pre_apply_query ... bench:         193 ns/iter (+/- 3)`

File: src/raftstore/store/store.rs [hunk 7]
```diff
@@ -189,6 +190,7 @@ impl<T, C> Store<T, C> {
         let mut coprocessor_host = CoprocessorHost::new();
         // TODO load coprocessors from configuration
         coprocessor_host.registry.register_observer(100, box SplitObserver);
+        coprocessor_host.registry.register_observer(99, box PrewriteObserver);
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: should define const name for 99 and 100 now.

File: Cargo.toml (hunk 2 above)
Comment (lishihai9017) [inline Cargo.toml]: not necessary, but it will covered later.

**Comment by lishihai9017** [comment]
PTAL again, @zhangjinpeng1987 @siddontang @BusyJay

**Comment by ngaut** [comment]
Please update bench result.

File: src/raftstore/coprocessor/txn_observer.rs [hunk 8]
```diff
@@ -24,47 +24,35 @@ impl Coprocessor for TxnObserver {}
...
-                reqs.push(req.to_owned());
+        for i in 0..reqs.len() {
+            //let ref mut req = reqs[i];
+            if reqs[i].get_cmd_type() != CmdType::Prewrite {
+                continue;
             }
+            let mut prewrite = reqs[i].take_prewrite();
+            let key = prewrite.take_key();
+            let lock_key = Key::from_encoded(key.clone())
+                .truncate_ts()
```
Comment (BusyJay) [inline src/raftstore/coprocessor/txn_observer.rs]: Can reduce allocation here too.

File: src/storage/mvcc/lock.rs [hunk 9]
```diff
@@ -56,7 +56,7 @@ impl LockType {
     }
 }
 
-#[derive(PartialEq, Debug)]
+#[derive(PartialEq, Clone, Debug)]
```
Comment (BusyJay) [inline src/storage/mvcc/lock.rs]: Why clone?

File: src/raftstore/store/store.rs [hunk 10]
```diff
@@ -73,6 +74,9 @@ type Key = Vec<u8>;
 const MIO_TICK_RATIO: u64 = 10;
 const PENDING_VOTES_CAP: usize = 20;
 
+const SPLIT_OBSERVER_IDX: u32   = 100;
+const TNX_OBSERVER_IDX: u32     = 99;
```
Comment (BusyJay) [inline src/raftstore/store/store.rs]: Remove extra spaces.

File: src/raftstore/store/worker/apply.rs [hunk 11]
```diff
@@ -886,7 +886,9 @@ impl ApplyDelegate {
                 CmdType::Put => self.handle_put(ctx, req),
                 CmdType::Delete => self.handle_delete(ctx, req),
                 CmdType::Snap => self.handle_snap(ctx, req),
-                CmdType::Invalid => Err(box_err!("invalid cmd type, message maybe currupted")),
+                CmdType::Invalid | CmdType::Prewrite => {
```
Comment (BusyJay) [inline src/raftstore/store/worker/apply.rs]: If Prewrite exists, there should be a bug somewhere.

File: src/storage/mvcc/txn.rs [hunk 12]
```diff
@@ -75,10 +75,23 @@ impl<'a> MvccTxn<'a> {
         self.writes.push(Modify::Delete(CF_LOCK, key));
     }
 
-    fn put_value(&mut self, key: &Key, ts: u64, value: Value) {
-        let key = key.append_ts(ts);
+    // fn put_value(&mut self, key: &Key, ts: u64, value: Value) {
```
Comment (BusyJay) [inline src/storage/mvcc/txn.rs]: Delete it instead of commenting out.

File: src/storage/mvcc/lock.rs (hunk 9 above)
Comment (lishihai9017) [inline src/storage/mvcc/lock.rs]: clone lock in TxnObserver test

**Pull request closed without merging**